File: backend/app/schemas/catalog.py
from typing import Literal

from pydantic import BaseModel, Field, HttpUrl

CurrencyCode = Literal["LKR", "USD"]
Availability = Literal["available", "unavailable"]
Category = Literal["oversized", "raglan", "customized"]
ImageRole = Literal["primary", "hover", "gallery"]


class OptionResponse(BaseModel):
    id: str
    label: str
    swatch: str | None = None


class PriceResponse(BaseModel):
    minorAmount: int
    currency: CurrencyCode


class ImageResponse(BaseModel):
    src: str
    alt: str
    width: int
    height: int


class ProductImagesResponse(BaseModel):
    primary: ImageResponse
    hover: ImageResponse | None = None
    gallery: list[ImageResponse]


class ThemeResponse(BaseModel):
    id: str
    label: str


class VariantResponse(BaseModel):
    id: str
    colorId: str
    sizeId: str
    availability: Availability
    prices: dict[CurrencyCode, PriceResponse] | None = None


class ProductResponse(BaseModel):
    id: str
    slug: str
    name: str
    category: Category
    garmentType: str
    theme: ThemeResponse | None
    description: str | None = None
    images: ProductImagesResponse
    colors: list[OptionResponse]
    sizes: list[OptionResponse]
    prices: dict[CurrencyCode, PriceResponse]
    variants: list[VariantResponse]
    availability: Availability


class ProductListResponse(BaseModel):
    items: list[ProductResponse]
    page: int
    pageSize: int
    total: int
    pages: int


class GarmentOptionsResponse(BaseModel):
    id: str
    label: str
    colors: list[OptionResponse]
    sizes: list[OptionResponse]
    standardPrices: dict[CurrencyCode, PriceResponse]
    customizedPrices: dict[CurrencyCode, PriceResponse]


class CatalogOptionsResponse(BaseModel):
    themes: list[ThemeResponse]
    garments: list[GarmentOptionsResponse]


class ProductImageCreate(BaseModel):
    storage_key: str | None = None
    public_url: HttpUrl | None = None
    alt_text: str = Field(min_length=1, max_length=300)
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    position: int = Field(ge=0)
    role: ImageRole


class ProductCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    slug: str = Field(pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$", max_length=160)
    description: str | None = None
    category: Category
    garment_slug: Literal["oversized", "raglan"]
    theme_slug: str | None = None
    color_slug: str
    size_slugs: list[str] = Field(min_length=1)
    images: list[ProductImageCreate] = Field(min_length=1)
    availability: Availability = "available"
    display_order: int = Field(default=0, ge=0)

File: backend/src/schemas/catalog.js
import { z } from "zod";

const image = z.object({
  media_asset_id: z.string().uuid().nullish(), storage_key: z.string().min(1).nullish(), public_url: z.url().nullish(),
  alt_text: z.string().trim().min(1).max(300), width: z.number().int().positive(), height: z.number().int().positive(),
  position: z.number().int().nonnegative(), role: z.enum(["primary", "hover", "gallery"]),
});
export const productInput = z.object({
  name: z.string().trim().min(1).max(200), slug: z.string().max(160).regex(/^[a-z0-9]+(?:-[a-z0-9]+)*$/),
  description: z.string().nullish(), category: z.enum(["oversized", "raglan", "customized"]),
  garment_slug: z.enum(["oversized", "raglan"]), theme_slug: z.string().nullish(), color_slug: z.string().min(1),
  size_slugs: z.array(z.string()).min(1), images: z.array(image).min(1),
  availability: z.enum(["available", "unavailable"]).default("available"), display_order: z.number().int().nonnegative().default(0),
});
export const productQuery = z.object({
  page: z.coerce.number().int().min(1).default(1), pageSize: z.coerce.number().int().min(1).max(100).default(24),
  garment: z.enum(["oversized", "raglan"]).optional(), theme: z.string().min(1).max(50).optional(),
  color: z.string().min(1).max(50).optional(), size: z.string().min(1).max(20).optional(),
  sort: z.enum(["display", "name-asc", "name-desc"]).default("display"),
});
